### ISSUE TITLE
Not a directory error on .png file on 2.6.0 RC2

### ISSUE THREAD

**Issue by aysiu** [munki/munki#585]
Weirdly got this on one client (four other clients working fine). The GateKeeper.png does exist and is not a directory. It also has the same permissions and ownership as the other .png files in the icons folder.

```
    **Processing self-serve choices**
    Icon base URL is: https://******************/icons/
    Options: {'logging_function': <function display_debug2 at 0x10e267578>, 'additional_headers': {u'Authorization': u'Basic *****************************', 'User-Agent': u'managedsoftwareupdate/2.6.0.2681 Darwin/14.5.0'}, 'file': u'/Library/Managed Installs/icons/GateKeeper.png.download', 'cache_data': None, 'url': u'https://*******************/icons/GateKeeper.png', 'follow_redirects': u'none', 'download_only_if_changed': False, 'can_resume': False}
2016-04-04 10:14:55.194 Python[1778:12093] *** Terminating app due to uncaught exception 'OC_PythonException', reason: '<type 'exceptions.IOError'>: [Errno 20] Not a directory: u'/Library/Managed Installs/icons/GateKeeper.png.download''
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff9448d03c __exceptionPreprocess + 172
	1   libobjc.A.dylib                     0x00007fff9b36376e objc_exception_throw + 43
	2   CoreFoundation                      0x00007fff9448cbd9 -[NSException raise] + 9
	3   _objc.so                            0x000000010b9dba8b PyObjCErr_ToObjCWithGILState + 46
	4   _objc.so                            0x000000010b9c852d PyObjCFFI_MakeClosure + 5097
	5   libffi.dylib                        0x00007fff9360ca07 ffi_closure_unix64_inner + 511
	6   libffi.dylib                        0x00007fff9360c0c6 ffi_closure_unix64 + 70
	7   Foundation                          0x00007fff9a6ba685 -[NSBlockOperation main] + 97
	8   Foundation                          0x00007fff9a6b9b3c -[__NSOperationInternal _start:] + 653
	9   Foundation                          0x00007fff9a6b9613 __NSOQSchedule_f + 184
	10  libdispatch.dylib                   0x00007fff8efade73 _dispatch_client_callout + 8
	11  libdispatch.dylib                   0x00007fff8efb15cd _dispatch_queue_drain + 1100
	12  libdispatch.dylib                   0x00007fff8efb1030 _dispatch_queue_invoke + 202
	13  libdispatch.dylib                   0x00007fff8efb0bef _dispatch_root_queue_drain + 463
	14  libdispatch.dylib                   0x00007fff8efb0a1c _dispatch_worker_thread3 + 91
	15  libsystem_pthread.dylib             0x00007fff9bd25a9d _pthread_wqthread + 729
	16  libsystem_pthread.dylib             0x00007fff9bd233dd start_wqthread + 13
)
libc++abi.dylib: terminating with uncaught exception of type NSException
Abort trap: 6

```

**Comment by aysiu** [comment]
It's not that particular icon. I removed that icon from the icons folder, and then it just choked on the next icon:

```
Getting icon GateKeeper.png for GateKeeper...
2016-04-04 10:25:07.283 Python[2515:16525] *** Terminating app due to uncaught exception 'OC_PythonException', reason: '<type 'exceptions.IOError'>: [Errno 20] Not a directory: u'/Library/Managed Installs/icons/Office2011_update.png.download''
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff9448d03c __exceptionPreprocess + 172
	1   libobjc.A.dylib                     0x00007fff9b36376e objc_exception_throw + 43
	2   CoreFoundation                      0x00007fff9448cbd9 -[NSException raise] + 9
	3   _objc.so                            0x000000010b913a8b PyObjCErr_ToObjCWithGILState + 46
	4   _objc.so                            0x000000010b90052d PyObjCFFI_MakeClosure + 5097
	5   libffi.dylib                        0x00007fff9360ca07 ffi_closure_unix64_inner + 511
	6   libffi.dylib                        0x00007fff9360c0c6 ffi_closure_unix64 + 70
	7   Foundation                          0x00007fff9a6ba685 -[NSBlockOperation main] + 97
	8   Foundation                          0x00007fff9a6b9b3c -[__NSOperationInternal _start:] + 653
	9   Foundation                          0x00007fff9a6b9613 __NSOQSchedule_f + 184
	10  libdispatch.dylib                   0x00007fff8efade73 _dispatch_client_callout + 8
	11  libdispatch.dylib                   0x00007fff8efb15cd _dispatch_queue_drain + 1100
	12  libdispatch.dylib                   0x00007fff8efb1030 _dispatch_queue_invoke + 202
	13  libdispatch.dylib                   0x00007fff8efb0bef _dispatch_root_queue_drain + 463
	14  libdispatch.dylib                   0x00007fff8efb0a1c _dispatch_worker_thread3 + 91
	15  libsystem_pthread.dylib             0x00007fff9bd25a9d _pthread_wqthread + 729
	16  libsystem_pthread.dylib             0x00007fff9bd233dd start_wqthread + 13
)
libc++abi.dylib: terminating with uncaught exception of type NSException
Abort trap: 6
```

**Comment by gregneagle** [comment]
And this is unique to v2.6.0RC2?  IOW, if you reinstall 2.5.x on the problematic client, the issue goes away?

**Comment by aysiu** [comment]
It had the previous client before and worked fine. I'll try that just to isolate the issue, though. Three other clients not complaining, so it may just be something's messed up there? Maybe the installer got corrupted in transfer?

I'll report back.

**Comment by gregneagle** [comment]
If you want other eyes on this, you should post something on munki-dev and/or munki-discuss. I'm not seeing anything like this here, and you and I are the only ones with eyes on this discussion.

**Comment by aysiu** [comment]
Yes. Confirmed. Reinstalled 2.5.1, and it works fine.

**Comment by aysiu** [comment]
I'll post on the lists. Thanks, Greg.

**Comment by gregneagle** [comment]
"Yes. Confirmed. Reinstalled 2.5.1, and it works fine.". Great. Now (re-)update to 2.6.0RC2. Problem returns?

And:
You seem to not notice that the paths it's choking on end with ".download":

`u'/Library/Managed Installs/icons/GateKeeper.png.download'`

Anything stand out there about those paths?

**Comment by aysiu** [comment]
Yeah, re-upgraded. Same problem.

I'm not sure about why it's looking for .png.download, since none of my other test clients are. In fact, I just loaded RC2 on to another client, and that client worked just fine. No issues whatsoever.

Anything in the logs that may be helpful besides the verbose output of
`/usr/local/munki/managedsoftwareupdate -vvv`
?

If it's just this one client, I may try nuking its Munki installation completely and reinstalling (including the preferences file and regenerated caches) from scratch in case there's something corrupt. But I want to leave it for now, in case it's not just this one client (i.e., other testers are reporting in, and we have to see what all these machines have in common).

**Comment by gregneagle** [comment]
It's not _looking_ for .png.download. It's downloading to a file named like /Library/Managed Installs/icons/FooBarBaz.png.download, then when the download is complete, renaming the .download file to the final name.

**Comment by aysiu** [comment]
That fixed it!

I was too busy looking at the server side and not noticing the error was actually pointing to the client side. For some strange reason, this machine had a zero-byte file named **/Library/Managed Installs/icons**. My other machines do not. I don't know how that file go there (or why 2.5.1 is okay with the file being there), but once I deleted the file, then `sudo /usr/local/munki/managedsoftwareupdate -vvv` worked just fine.

**Comment by gregneagle** [comment]
Could you be more specific on what the actual problem was and how you fixed it?

**Comment by gregneagle** [comment]
(Sounds like maybe it was trying to create a file in /Library/Managed Installs/icons, but that path was not a directory; instead it was a _file_.)

**Comment by aysiu** [comment]
> Could you be more specific on what the actual problem was and how you fixed it?

I'm not really sure what the problem was. I did a
```ls -al /Library/Managed\ Installs```
on both the non-working client and on a working client.

On the non-working client, I saw this line:
```-rw-r--r--    1 root  admin          0 Apr  4 10:36 icons```

On the working client, I saw no such line. Nothing called _icons_ at all in that directory.

So I just did a
```sudo rm /Library/Managed\ Installs/icons```
on the non-working client, and then it worked after that.

Maybe I'm not understanding your question, but if there are other details I can provide that would be helpful, let me know.
